File: 1996-number-of-ways-to-rearrange-sticks-with-k-sticks-visible/number-of-ways-to-rearrange-sticks-with-k-sticks-visible.c
#define MOD 1000000007

int rearrangeSticks(int n, int k) {
    long long dp[k + 1];

    for (int i = 0; i <= k; i++)
        dp[i] = 0;

    dp[0] = 1;

    for (int i = 1; i <= n; i++) {
        int limit = (i < k) ? i : k;

        for (int j = limit; j >= 1; j--) {
            dp[j] = (dp[j - 1] + (long long)(i - 1) * dp[j]) % MOD;
        }

        dp[0] = 0;
    }

    return (int)dp[k];
}

File: 1996-number-of-ways-to-rearrange-sticks-with-k-sticks-visible/number-of-ways-to-rearrange-sticks-with-k-sticks-visible.kt
class Solution {
    fun rearrangeSticks(n: Int, k: Int): Int {
        val MOD = 1_000_000_007L
        val dp = LongArray(k + 1)
        dp[0] = 1L

        for (i in 1..n) {
            val limit = minOf(i, k)
            for (j in limit downTo 1) {
                dp[j] = (dp[j - 1] + (i - 1).toLong() * dp[j]) % MOD
            }
            dp[0] = 0L
        }

        return dp[k].toInt()
    }
}